Improvements to `PhasingViaCostFunction` infrastructure

tanujkhattar opened the pull request:
Fixes https://github.com/quantumlib/Qualtran/issues/654

fdmalone commented on qualtran/bloqs/rotations/phasing_via_cost_function_test.py:
```diff
@@ -75,7 +77,7 @@ def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str
     n = 2
     exponent = 1.20345
-    eps = 5e-4
+    eps = 1e-3
```
Why did this change?

tanujkhattar replied:
to make the test run quickly

fdmalone commented on qualtran/bloqs/rotations/phase_gradient.py:
```diff
@@ -149,26 +147,46 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
-    inp_bitsize: int
+    x_bitsize: int
     phase_bitsize: int
+    right_shift: int = 0
```
update docstring.

tanujkhattar replied:
done

fdmalone commented on qualtran/bloqs/rotations/phase_gradient.py:
```diff
@@ -135,10 +135,8 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
         U|x\rangle|\text{phase\_grad}\rangle = |x\rangle|\text{phase\_grad} + x\rangle
     $$
 
-    TODO(#654): Check whether phase_bitsize >= inp_bitsize needs to be enforced.
```
Should probably capture the conclusion of this todo in the args somewhere.

tanujkhattar replied:
See the newly added and very detailed `qualtran/bloqs/rotations/quantum_variable_rotation.ipynb` notebook.

fdmalone commented on qualtran/bloqs/rotations/phase_gradient.py:
```diff
@@ -149,26 +147,46 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
+    def scaled_val(self, x: int) -> int:
+        """Computes `phase_grad + x` using fixed point arithmetic."""
+        x_width = self.x_bitsize + self.right_shift
+        x_fxp = _fxp(x / 2**x_width, x_width).like(_fxp(0, self.phase_bitsize)).astype(float)
```
am i reading this line right:

1. x / 2**x_width of width x_width 
2. Convert to a new fxp of width phase_bitsize
3. convert to float.

Why is step 2 required?

fdmalone commented on qualtran/bloqs/rotations/phase_gradient.py:
```diff
@@ -149,26 +147,46 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
 
-    def on_classical_vals(self, x, phase_grad) -> Dict[str, 'ClassicalValT']:
-        return {'x': x, 'phase_grad': (phase_grad + x) % (2**self.phase_bitsize)}
+    def on_classical_vals(self, x: int, phase_grad: int) -> Dict[str, 'ClassicalValT']:
```
I'm a bit confused, so on input x should be the the integer representation of what we want our fixed point real number to be. What values can x take?

tanujkhattar replied:
Ideally input `x` would be an `Fxp` object instead of `int` but we don't have support for quantum data types in classical simulation yet. I've added more details in the newly added notebook `qualtran/bloqs/rotations/quantum_variable_rotation.ipynb`

fdmalone commented on qualtran/bloqs/rotations/phase_gradient.py:
```diff
@@ -179,6 +197,33 @@ def _t_complexity_(self) -> 'TComplexity':
 
 
+def _fxp(x: float, out: int):
+    assert 0 <= x < 1
```
maybe add a little context here so we don't have to google the Fxp kwargs.

tanujkhattar replied:
done

fdmalone commented on qualtran/bloqs/rotations/phase_gradient.py:
```diff
@@ -237,12 +290,11 @@ def scaled_val(self, x: int) -> int:
         # However, `x` should be interpreted as per the fixed point specification given in self.inp_dtype.
```
inp_dtype is no more.

fdmalone commented on qualtran/bloqs/rotations/phase_gradient.py:
```diff
@@ -224,11 +269,19 @@ def with_registers(self, *new_registers: Union[int, Sequence[int]]):
         n_int = Fxp(abs(self.gamma), signed=False).n_int
         return QFxp(n_int + self.gamma_bitsize, self.gamma_bitsize, signed=False)
 
+    @cached_property
```
Why do we take abs(gamma) and not just keep it signed?

tanujkhattar replied:
since `gamma` is a classically known constant, we can decide at the time of circuit construction whether we wish to add or subtract the scaled value into the target register. Thus, we don't need to waste one bit of gamma to store the sign information. This is why `AddIntoPhaseGrad` accepts a `sign: int` which is essentially the `sign(gamma)`

fdmalone commented on qualtran/bloqs/rotations/phasing_via_cost_function.py:
```diff
@@ -154,30 +158,20 @@ def b_phase(self) -> int:
-        #   `cost_reg.bitsize`.
-        assert eq_a7 >= eq_35
-        return eq_35
+        return int(np.ceil(np.log2((self.gamma_bitsize + 2) * np.pi / self.eps)))
```
What was the resolution here? A7 is correct?

tanujkhattar replied:
See the newly added notebook `qualtran/bloqs/rotations/quantum_variable_rotation.ipynb`. 

Tl;Dr is that both equations were correct and equivalent given the assumptions in the paper. In the general case, A7 is more accurate.

fdmalone commented on qualtran/bloqs/rotations/phasing_via_cost_function.py:
```diff
@@ -154,30 +158,20 @@ def b_phase(self) -> int:
+        # Using `gamma_bitsize = log(gamma) + b_{phase} + O(1)` defined b/w equation 34 & 35
+        # of https://arxiv.org/abs/2007.07391. Note that the `log(gamma)` here means ignoring
+        # leading digits of `gamma` which are 0
+        return self.b_phase
```
We can get away with the + O(1) here? Should this be determined through some simulation?

tanujkhattar replied:
The + O(1) is due log(2pi); updated the implementation and added lots of details in the new notebook.

fdmalone reviewed (changes requested):
I just left several questions. Maybe worth using this time to add the autodoc examples to capture the nice docstrings.

mpharrigan commented on qualtran/bloqs/rotations/hamming_weight_phasing.py:
```diff
@@ -168,7 +175,10 @@ def build_composite_bloq(
         out, phase_grad = bb.add(
             AddScaledValIntoPhaseReg(
-                self.phase_oracle.cost_dtype, self.b_grad, self.exponent / 2, self.gamma_bitsize
+                self.phase_oracle.cost_dtype.bitsize,
+                self.b_grad,
+                (self.exponent / 2) * (2 ** self.bitsize.bit_length()),
+                self.gamma_bitsize,
```
keyword arguments would not be amiss here

tanujkhattar replied:
added

mpharrigan commented:
+1 to @fdmalone 's comments about recording the result of this investigation in the codebase

tanujkhattar commented:
See qualtran/bloqs/rotations/quantum_variable_rotation.ipynb for a very detailed pedagogical review (which I think I've done better than the linked references, and also explained a more general case) of all the details behind QVR via phase gradients. 

I've also added LOTS of more docstrings and auto generated notebooks. Please take another look. 

Since the size of the PR has grown significantly and there are a lot of nuanced details; I'd like to get this merged soon to avoid merge conflicts with anything else. I'll present the detailed review in the Qualtran sync tomorrow as well.

fdmalone commented on qualtran/bloqs/rotations/quantum_variable_rotation.ipynb:
nit: the *'s here should probably be \times to render a bit nicer.

fdmalone commented on qualtran/bloqs/rotations/quantum_variable_rotation.ipynb:
nit: >= -> \ge

tanujkhattar replied:
done

mpharrigan commented on qualtran/bloqs/rotations/quantum_variable_rotation.py:
```diff
@@ -0,0 +1,339 @@
+
+    A $b_\text{grad}$-bit phase gradient state $|\phi\rangle_{b_\text{grad}}$ can be written as
+
+    <!-- can be thought of as a superposition of states $|\frac{0}{2^{b_\text{grad}}}\rangle, |\frac{2}{2^{b_\text{grad}}}\rangle, |\frac{2}{2^{b_\text{grad}}}\rangle, ... |k/$ where each integer $0\leq k \lt 2^{b_\text{grad}}$ represents a $b_\text{grad}$-bit fixed point number $\frac{k}{2^{b_\text{grad}}}$ -->
```
because of the line length? yeah, curious.

What's with the html-style comments? are these omitted from the notebook?

mpharrigan replied:
all of the "C" lint  check s are disabled including `line-too-long` since the formatter usually handles it.

mpharrigan replied:
but please wrap these docstrings. Your IDE should be able to do most of it for you

fdmalone replied:
black can't handle docstrings right

tanujkhattar replied:
wrapped the docstrings, removed html comments (leftover by mistake)

tanujkhattar commented:
I've updated the docstrings and added some more tests to `quantum_variable_rotation_test.py`. I'm unable to prove the right precision for the case when `b_grad < x_bitsize` but I believe the precision should depend only upon `log(1/eps)` and not scale with size of the cost register. I've update the docstrings to include this information, added some worst case tests where `b_grad < x_bitsize`.  I'll merge this PR now and we can revisit later if we discover this assumption to be incorrect.
